plonky2/src/hash/merkle_tree.rs

```diff
@@ -27,36 +30,108 @@ pub struct MerkleTree<F: RichField, H: Hasher<F>> {
     /// The data in the leaves of the Merkle tree.
     pub leaves: Vec<Vec<F>>,
 
-    /// The layers of hashes in the tree. The first layer is the one at the bottom.
-    pub layers: Vec<Vec<H::Hash>>,
+    /// The digests in the tree. Consists of `cap.len()` sub-trees, each corresponding to one
+    /// element in `cap`. Each subtree is contiguous and located at
+    /// `digests[digests.len() / cap.len() * i..digests.len() / cap.len() * (i + 1)]`.
+    /// Within each subtree, siblings are stored next to each other. The layout is,
+    /// left_child_subtree || left_child_digest || right_child_digest || right_child_subtree, where
+    /// left_child_digest and right_child_digest are H::Hash and left_child_subtree and
+    /// right_child_subtree recurse. Observe that the digest of a node is stored by its _parent_.
+    /// Consequently, the digests of the roots are not stored here (they can be found in `cap`).
+    pub digests: Vec<H::Hash>,
 
     /// The Merkle cap.
     pub cap: MerkleCap<F, H>,
 }
 
+fn capacity_up_to_mut<T>(v: &mut Vec<T>, len: usize) -> &mut [MaybeUninit<T>] {
+    assert!(v.capacity() >= len);
+    let v_ptr = v.as_mut_ptr().cast::<MaybeUninit<T>>();
+    unsafe {
+        // SAFETY: `v_ptr` is a valid pointer to a buffer of length at least `len`. Upon return, the
+        // lifetime will be bound to that of `v`. The underlying memory will not be deallocated as
+        // we hold the sole mutable reference to `v`. The contents of the slice may be
+        // uninitialized, but the `MaybeUninit` makes it safe.
+        slice::from_raw_parts_mut(v_ptr, len)
+    }
+}
+
+fn fill_subtree<F: RichField, H: Hasher<F>>(
+    digests_buf: &mut [MaybeUninit<H::Hash>],
+    leaves: &[Vec<F>],
+) -> H::Hash {
+    assert_eq!(leaves.len(), digests_buf.len() / 2 + 1);
+    if digests_buf.is_empty() {
+        H::hash(&leaves[0], false)
+    } else {
+        // Layout is: left recursive output || left child digest
+        //             || right child digest || right recursive output.
+        // Split `digests_buf` into the two recursive outputs (slices) and two child digests
+        // (references).
+        let (left_digests_buf, right_digests_buf) = digests_buf.split_at_mut(digests_buf.len() / 2);
+        let (left_digest_mem, left_digests_buf) = left_digests_buf.split_last_mut().unwrap();
+        let (right_digest_mem, right_digests_buf) = right_digests_buf.split_first_mut().unwrap();
+        // Split `leaves` between both children.
+        let (left_leaves, right_leaves) = leaves.split_at(leaves.len() / 2);
+        let (left_digest, right_digest) = rayon::join(
```

Comment:
I thought this might be a bit much overhead, spawning a task for each hash. It's easy to adjust the granularity with e.g.

```rust
        let mut fill_left = || fill_subtree::<F, H>(left_digests_buf, left_leaves);
        let mut fill_right = || fill_subtree::<F, H>(right_digests_buf, right_leaves);
        let (left_digest, right_digest) = if leaves.len() > 16 {
            rayon::join(fill_left, fill_right)
        } else {
            (fill_left(), fill_right())
        };
```

But I couldn't really detect a change in performance. Might be worth trying on another machine but I guess the overhead is small.

Reply:
[The docs](https://docs.rs/rayon/latest/rayon/fn.join.html) claim that `rayon::join` is actually more clever than blindly spinning up tasks. In summary, the calling worker thread executes the first task, but the second task is marked as “available for stealing” if another thread is looking for work in the meantime. If it doesn’t get stolen (as we would expect the vast majority of the time), then the calling thread proceeds to execute the second task.

You’re still right though that this adds _some_ overhead—just the atomic writes involved would not be cheap—but Poseidon (at ~2K per hash and many hashes per leaf) would dwarf that.